core/templates/dev/head/player/AnswerClassificationServiceSpec.js

```diff
@@ -0,0 +1,137 @@
+// Copyright 2015 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for the answer classification service
+ *
+ * @author wxyxinyu@gmail.com (Xinyu Wu)
+ */
+
+describe('Answer classifiation service', function() {
+  beforeEach(module('oppia'));
+
+  var acs, $httpBackend;
+  beforeEach(inject(function($injector) {
+    acs = $injector.get('answerClassificationService');
+    $httpBackend = $injector.get('$httpBackend');
+  }));
+
+  var explorationId = 'exploration';
+
+  var state = {
+    interaction: {
+      answer_groups: [{
+        outcome: 'outcome 1',
+        rule_specs: [{
+          inputs: {
+            x: 10
+          },
+          rule_type: 'Equals'
+        }]
+      }, {
+        outcome: 'outcome 2',
+        rule_specs: [{
+          inputs: {
+            x: 5
+          },
+          rule_type: 'Equals'
+        }, {
+          inputs: {
+            x: 7
+          },
+          rule_type: 'NotEquals'
+        }]
+      }],
+      default_outcome: 'default'
+    }
+  };
+
+  var rules = {
+    'Equals': function(answer, inputs) {
+      return inputs.x === answer;
+    },
+    'NotEquals': function(answer, inputs) {
+      return inputs.x != answer;
+    }
+  };
+
+  it('should query the backend when no frontend rules are provided',
+      function() {
+    var backendClassifiedOutcome = {
+      outcome: 'outcome',
+      answer_group_index: 0,
+      rule_spec_index: 0
+    };
+    var expectedClassificationResult = {
+      outcome: 'outcome',
+      answerGroupIndex: 0,
+      ruleSpecIndex: 0
+    };
+    $httpBackend.expectPOST(
+      '/explorehandler/classify/' + explorationId).respond(
+      backendClassifiedOutcome);
+    var classificationResult;
+    acs.getMatchingClassificationResult(explorationId, state, 0, false).
+      then(function(outcome) {
+        classificationResult = outcome;
+    });
+    $httpBackend.flush();
+    expect(classificationResult).toEqual(expectedClassificationResult);
+    $httpBackend.verifyNoOutstandingExpectation();
+    $httpBackend.verifyNoOutstandingRequest();
+  });
+
+  it('should return the first matching answer group and first matching rule' +
+     'spec', function() {
+    acs.getMatchingClassificationResult(
+      explorationId, state, 10, false, rules).then(
```

Comment:
I think we want to use jasmine.createSpy() as a pattern instead, since it will not only verify the input of the resolve function but will also assert the promise is called at all. For an example of what this looks like, see:

https://github.com/oppia/oppia/blob/d977c6bcb7243dd2959f2177fd363a540dafbdf9/core/templates/dev/head/domain/collection/CollectionDataServiceSpec.js#L68